File: matlab/tests/bench_orbit_xlang.m
function bench_orbit_xlang
%BENCH_ORBIT_XLANG  Cross-language speed benchmark for orbit IP/eval.
%
%   Targeted spot-checks (not a sweep) at the same (r, K, P)
%   configurations as the Python companion
%   ``python/tests/bench_orbit_xlang.py``. Prints CSV-format lines
%   matching the Python output so the two can be concatenated and
%   read into a comparison table.
%
%   Compare median wall times to gauge whether MATLAB is within ~2x
%   of Python (the gating threshold for shipping the v2.2 MATLAB port
%   as-is vs investing in precomputed contraction paths or a sharper
%   greedy heuristic in mobius.contract).
%
%   Run from the matlab/ directory after addpath('+mobius'):
%     >> bench_orbit_xlang
%   then redirect or copy-paste the output and concatenate with the
%   Python output for analysis.

    % --- Configuration (must match Python script) ---
    configs = struct( ...
        'r', {2, 3, 3, 4, 5}, ...
        'K', {8, 8, 12, 8, 6}, ...
        'P', {50, 50, 50, 20, 10});

    N_REPS    = 7;
    N_WARMUP  = 2;

    fprintf('language,bench,r,K,P,t_median_ms\n');
    for ci = 1:numel(configs)
        r = configs(ci).r;
        K = configs(ci).K;
        P = configs(ci).P;

        t_single  = benchOrbitSingle(r, K, N_REPS, N_WARMUP) * 1000;
        t_batched = benchOrbitBatched(r, K, P, N_REPS, N_WARMUP) * 1000;
        t_cossim  = benchCosSimSingleMultisetOrbit(r, K, N_REPS, N_WARMUP) * 1000;

        fprintf('matlab,inner_product_orbit,%d,%d,1,%.4f\n', ...
                r, K, t_single);
        fprintf('matlab,inner_product_orbit_pw_batched,%d,%d,%d,%.4f\n', ...
                r, K, P, t_batched);
        fprintf('matlab,cos_sim_exp_tens_sa_orbit,%d,%d,1,%.4f\n', ...
                r, K, t_cossim);
    end
end


% =========================================================================
%  Helpers
% =========================================================================

function [K, w_a, w_b] = makeKernel(K_x, K_y, seed)
    rng(seed, 'twister');
    p_x = sort(2000 * rand(K_x, 1));
    p_y = sort(2000 * rand(K_y, 1));
    sigma = 30;
    K = exp(-(p_x - p_y.').^2 / (4 * sigma^2));
    w_a = 0.5 + rand(K_x, 1);
    w_b = 0.5 + rand(K_y, 1);
end


function t = timed(fn, nReps, nWarmup)
    for w = 1:nWarmup
        fn();
    end
    times = zeros(nReps, 1);
    for i = 1:nReps
        t0 = tic;
        fn();
        times(i) = toc(t0);
    end
    t = median(times);
end


% =========================================================================
%  Benchmarks
% =========================================================================

function t = benchOrbitSingle(r, K, nReps, nWarmup)
    [K_mat, w_a, w_b] = makeKernel(K, K, r * 7);
    t = timed(@() mobius.innerProductOrbit(K_mat, w_a, w_b, r, ...
                                              'prefactor', 1.0), ...
              nReps, nWarmup);
end


function t = benchOrbitBatched(r, K, P, nReps, nWarmup)
    K_pairs = zeros(P, K, K);
    w_A = zeros(P, K);
    w_B = zeros(P, K);
    for i = 1:P
        [K_mat, w_a, w_b] = makeKernel(K, K, r * 11 + i);
        K_pairs(i, :, :) = K_mat;
        w_A(i, :) = w_a;
        w_B(i, :) = w_b;
    end
    t = timed(@() mobius.innerProductOrbitPwBatched( ...
                K_pairs, w_A, w_B, r, 'prefactor', 1.0), ...
              nReps, nWarmup);
end


function t = benchCosSimSingleMultisetOrbit(r, K, nReps, nWarmup)
    rng(r * 13 + K, 'twister');
    p1 = sort(2000 * rand(K, 1));
    p2 = sort(2000 * rand(K, 1));
    w  = ones(K, 1);
    t = timed(@() cosSimExpTens(p1, w, p2, w, 30, r, false, false, 0, ...
                                  'method', 'mobius', 'verbose', false), ...
              nReps, nWarmup);
end

File: python/tests/bench_orbit_xlang.py
"""Cross-language speed benchmark for the orbit IP/eval machinery.

Targeted spot-checks (not a sweep) at three representative (r, K)
configurations. The companion MATLAB script
``matlab/tests/bench_orbit_xlang.m`` runs the same configurations.
Compare median wall times to gauge whether MATLAB is within ~2x of
Python (the gating threshold for shipping the v2.2 MATLAB port as-is
vs investing in precomputed contraction paths or a sharper greedy
heuristic in ``mobius.contract``).

What we measure:
  - ``inner_product_orbit`` on a single (K_x, K_y) kernel.
  - ``inner_product_orbit_pw_batched`` on a batch of P pairs.
  - End-to-end ``cos_sim_exp_tens`` single-multiset orbit (a sanity check that
    includes dispatcher overhead).

Each configuration runs a small warm-up to amortise import / cache
costs, then reports the median over N_REPS runs.

Output format is fixed (one CSV-style line per measurement) so the
MATLAB script can produce a directly comparable file. Concatenate
both files and read into a comparison table.
"""

import time
import numpy as np

import mpt
from mpt._mobius import inner_product_orbit, inner_product_orbit_pw_batched
from mpt.tensor import cos_sim_exp_tens


# ----------------------------------------------------------------------
# Configuration
# ----------------------------------------------------------------------

# (r, K) configurations to spot-check. Span the typical range:
# r=2..5, K small enough to keep runtimes < 10 s/each per MATLAB run.
CONFIGS = [
    {'r': 2, 'K': 8,  'P': 50},
    {'r': 3, 'K': 8,  'P': 50},
    {'r': 3, 'K': 12, 'P': 50},
    {'r': 4, 'K': 8,  'P': 20},
    {'r': 5, 'K': 6,  'P': 10},
]

N_REPS = 7         # median of N_REPS timings per measurement
N_WARMUP = 2       # untimed warmup runs


# ----------------------------------------------------------------------
# Helpers
# ----------------------------------------------------------------------

def _make_kernel(K_x, K_y, seed=0):
    """Reproducible (K_x, K_y) Gaussian-style kernel + weight vectors."""
    rng = np.random.default_rng(seed)
    p_x = np.sort(rng.uniform(0, 2000, K_x))
    p_y = np.sort(rng.uniform(0, 2000, K_y))
    sigma = 30.0
    K = np.exp(-(p_x[:, None] - p_y[None, :]) ** 2 / (4 * sigma ** 2))
    w_a = rng.uniform(0.5, 1.5, K_x)
    w_b = rng.uniform(0.5, 1.5, K_y)
    return K, w_a, w_b


def _timed(fn, n_warmup=N_WARMUP, n_reps=N_REPS):
    """Median wall-clock seconds over n_reps runs (after n_warmup)."""
    for _ in range(n_warmup):
        fn()
    times = []
    for _ in range(n_reps):
        t0 = time.perf_counter()
        fn()
        times.append(time.perf_counter() - t0)
    return float(np.median(times))


# ----------------------------------------------------------------------
# Benchmarks
# ----------------------------------------------------------------------

def bench_orbit_single(r, K, P_unused):
    """Time inner_product_orbit on a single (K, K) kernel."""
    K_mat, w_a, w_b = _make_kernel(K, K, seed=r * 7)
    return _timed(
        lambda: inner_product_orbit(K_mat, w_a, w_b, r, prefactor=1.0),
    )


def bench_orbit_batched(r, K, P):
    """Time inner_product_orbit_pw_batched on P pairs of (K, K) kernels."""
    rng = np.random.default_rng(seed=r * 11 + K)
    K_pairs = np.empty((P, K, K))
    w_A = np.empty((P, K))
    w_B = np.empty((P, K))
    for i in range(P):
        K_mat, w_a, w_b = _make_kernel(K, K, seed=r * 11 + i)
        K_pairs[i] = K_mat
        w_A[i] = w_a
        w_B[i] = w_b
    return _timed(
        lambda: inner_product_orbit_pw_batched(
            K_pairs, w_A, w_B, r, prefactor=1.0,
        ),
    )


def bench_cossim_orbit_sm(r, K, P_unused):
    """End-to-end cos_sim_exp_tens single-multiset orbit (with dispatcher overhead)."""
    rng = np.random.default_rng(seed=r * 13 + K)
    p1 = np.sort(rng.uniform(0, 2000, K))
    p2 = np.sort(rng.uniform(0, 2000, K))
    w = np.ones(K)
    return _timed(
        lambda: cos_sim_exp_tens(
            p1, w, p2, w, 30.0, r, False, False, 0.0,
            method='mobius', verbose=False,
        ),
    )


# ----------------------------------------------------------------------
# Main
# ----------------------------------------------------------------------

if __name__ == '__main__':
    print("language,bench,r,K,P,t_median_ms")
    for cfg in CONFIGS:
        r, K, P = cfg['r'], cfg['K'], cfg['P']
        # Skip r=2,K=8: orbit at r=2 has only 4 orbit terms; both
        # languages are sub-millisecond and the comparison is noisy.
        # Keep it as a sanity floor anyway.

        t_single  = bench_orbit_single(r, K, P) * 1000
        t_batched = bench_orbit_batched(r, K, P) * 1000
        t_cossim  = bench_cossim_orbit_sm(r, K, P) * 1000

        print(f"python,inner_product_orbit,{r},{K},1,{t_single:.4f}")
        print(f"python,inner_product_orbit_pw_batched,{r},{K},{P},{t_batched:.4f}")
        print(f"python,cos_sim_exp_tens_sa_orbit,{r},{K},1,{t_cossim:.4f}")
